libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt

```diff
@@ -266,6 +271,146 @@ internal class NativeLoginManager(
+        val startRegistrationResponse = suspendedRestCall(startRegistrationRequest) ?: return StartRegistrationResult(UnknownError)
+
+        // React to the start registration response.
+        return when (startRegistrationResponse.isSuccess) {
+            true -> {
+                startRegistrationResponse.consumeQuietly()
+                startRegistrationResponse.asJSONObject().let { response ->
+                    StartRegistrationResult(
+                        nativeLoginResult = Success,
+                        email = response.get("email").toString(),
+                        requestIdentifier = response.get("identifier").toString()
+                    )
+                }
+            }
+
+            else -> {
+                SalesforceSDKLogger.e(
+                    TAG,
+                    "Cannot start registration as the Salesforce Identity API returned an unsuccessful HTTP status. '${startRegistrationResponse.asString()}'"
+                )
+                StartRegistrationResult(UnknownError)
+            }
+        }
+    }
+
+    override suspend fun completeRegistration(
+        otp: String,
+        requestIdentifier: String,
+        otpVerificationMethod: OtpVerificationMethod
+    ) = submitAuthorizationRequest(
+        authRequestType = AUTH_REQUEST_TYPE_VALUE_USER_REGISTRATION,
+        otp = otp,
+        identifier = requestIdentifier,
+        otpVerificationMethod = otpVerificationMethod
+    )
+
+    /**
+     * A data class for the Salesforce Identity API start registration request
+     * body.
+     * @param recaptcha The reCAPTCHA token provided by the reCAPTCHA Android
+     * SDK.  This is not used with reCAPTCHA Enterprise
+     * @param recaptchaevent The reCAPTCHA parameters for use with reCAPTCHA
+     * Enterprise
+     * @param userData The start registration request user data
+     * @param otpVerificationMethod The one-time-password's delivery method for
+     * verification in "email" or "sms"
+     */
+    private data class StartRegistrationRequestBody(
+        val recaptcha: String?,
+        val recaptchaevent: ReCaptchaEventRequestParameter?,
+        val userData: UserData,
+        val otpVerificationMethod: String
+    ) {
+        fun toJson() = JSONObject().apply {
+            put("recaptcha", recaptcha)
+            put("recaptchaevent", recaptchaevent?.toJson())
+            put("userdata", userData.toJson())
+            put("verificationmethod", otpVerificationMethod)
+        }
+
+        /**
+         * A data class for the Salesforce Identity API start registration
+         * request body's user info parameter.
+         * @param email A valid, user-entered email address
+         * @param username A valid Salesforce username or email
+         * @param password The user-entered new password
+         * @param firstName The user-entered first name
+         * @param lastName The user-entered last name
+         */
+        data class UserData(
```

Comment:
I subtle difference here from previous methods is that we have an object type as one of the fields, so here's that type.